eos-grub-upgrade

```diff
@@ -0,0 +1,276 @@
+#!/bin/bash
+# Copyright (C) 2018 Endless Mobile, Inc.
+# Licensed under the GPLv2
+#
+# Extracted from https://phabricator.endlessm.com/w/software/x86-64-boot-paths/
+#
+# +------+------------------------------------------+---------------+---------------------------------------------------------------+
+# | Type | Install type                             | Firmware / PT | GRUB / SHIM location                                          |
+# +------+------------------------------------------+---------------+---------------------------------------------------------------+
+# | A    | Standalone, since 3.1.0                  | EFI-64 / GPT  | GRUB_AMD64: \EFI\Endless\grubx64.efi in ESP on same disk as / |
+# |      |                                          |               | GRUB_IA32:  \EFI\BOOT\bootia32.efi in ESP                     |
+# |      |                                          |               | SHIM:       \EFI\BOOT\bootx64.efi and \EFI\endless\shim.efi   |
+# +------+------------------------------------------+---------------+---------------------------------------------------------------+
+# | B    | Standalone, since 3.0.0 and before 3.1.0 | EFI-64 / GPT  | GRUB_AMD64: \EFI\BOOT\grubx64.efi in ESP on same disk as /    |
+# |      |                                          |               | GRUB_IA32:  none                                              |
+# |      |                                          |               | SHIM:       \EFI\BOOT\bootx64.efi                             |
+# +------+------------------------------------------+---------------+---------------------------------------------------------------+
+# | C    | Standalone, before 3.0.0                 | EFI-64 / GPT  | GRUB:       \EFI\BOOT\bootx64.efi in ESP on same disk as /    |
+# |      |                                          |               | GRUB_IA32:  none                                              |
+# |      |                                          |               | SHIM:       none                                              |
+# +------+------------------------------------------+---------------+---------------------------------------------------------------+
+
+ESP_MOUNTPOINT=/boot/efi
+
+##########################################################################
+## UEFI
+##########################################################################
+
+## AMD64 GRUB
+declare -A GRUB_AMD64=(
+  [A]="${ESP_MOUNTPOINT}/EFI/endless/grubx64.efi"
+  [B]="${ESP_MOUNTPOINT}/EFI/BOOT/grubx64.efi"
+  [C]="${ESP_MOUNTPOINT}/EFI/BOOT/bootx64.efi"
+)
+NEW_GRUB_AMD64=/usr/lib/grub/images/x86_64-efi/grubx64.efi
+
+## SHIM
+declare -A SHIM=(
+  [A]="${ESP_MOUNTPOINT}/EFI/BOOT/bootx64.efi"
```

Comment:
So I see that what's happening here is that A, B and C are arbitrary labels, and the data structure you're simulating is something like:

```js
AMD64_LAYOUTS = [
    {grub: '/EFI/endless/grubx64.efi', shims: ['/EFI/BOOT/bootx64.efi', '/EFI/Endless/shim.efi']},
    // ...
]
```

It's a shame to have to disguise this structure as a bunch of strings and string->string maps, complete with runtime-evaluated variable names…